### PR TITLE
[IMP] survey: add background image on survey sections - alternative

### PR DESCRIPTION
**Pull request by dbeguin**
This merge adds the background per section feature.
If the section to display has a background image configured, the background
will be refreshed using that section background image.
If the section has no background, the survey background is used.
The background is refreshed only if the previous and the next sections are
different. The background is refreshed at the same time that the next question
is loaded.

To ease technical maintenance and to keep it simple, the background is always
faded out on back or if survey has conditional questions. This is to avoid to
load all the survey structure and to copy the get_next_question logic in the
frontend. When loading the next question, we check that the question that will
follow afterwards has the same background. If not, we force the background fade
out when user goes to the next question (refresh_background_on_next).
Otherwize, we only fade out the question and not the background to keep a
smooth transition between questions and/or sections.

Note: The next section to display depends on free text (section with
description) configuration and on conditional questions.
The next section to display can be the next question's section or directly
the next question itself (if the question is a section).

Task ID: 2225393

### CODE REVIEW

File: addons/survey/controllers/survey_session_manage.py [hunk 1]
```diff
@@ -110,14 +111,32 @@ def survey_session_next_question(self, survey_token, go_back=False, **kwargs):
...
+            template_values.update({
+                'background_image_url': background_image_url,
+                'reload_background_on_next': reload_background_on_next,
+            })
+
+            return {
+                'background_image_url': background_image_url,
+                # 'reload_background_on_next': reload_background_on_next,
+                'question_html': request.env.ref('survey.user_input_session_manage_content')._render(template_values)
+            }
```
Comment (awa-odoo) [inline addons/survey/controllers/survey_session_manage.py:137]: As this is only done here (when AJAX submitting the question), and not in the "survey_session_manage" route, I'm wondering if it ill work if the survey host makes a refresh of the page (F5) mid-session?

File: addons/survey/static/src/js/survey_form.js [hunk 2]
```diff
@@ -341,7 +341,9 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
                     route: `/survey/next_question/${this.options.surveyToken}/${this.options.answerToken}`,
                 }), {
                     initTimer: true,
-                    isFinish: nextPageEvent.type === 'end_session'
+                    isFinish: nextPageEvent.type === 'end_session',
+                    isNotification: true,
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_form.js:345]: What does the "isNotification" key do exactly?

File: addons/survey/static/src/js/survey_form.js [hunk 3]
```diff
@@ -430,9 +433,31 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
...
+        var refreshBackground = self.options.reloadBackgroundOnNext || options.goBack;
+        if (refreshBackground) {
+            // If we make a transition, hide previous background by setting 'background overlay' opacity = 1
+            $('div#wrapwrap').css("box-shadow", "inset 0 0 0 10000px rgba(255,255,255,1)");
+        }
 
         Promise.all([fadeOutPromise, nextScreenPromise]).then(function (results) {
-            return self._onNextScreenDone(results[1], options);
+            var refreshBackgroundPromise = Promise.resolve(false);
+            var next_question_background = results[1].background_image_url;
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_form.js:445]: camelCase?

File: addons/survey/static/src/js/survey_form.js [hunk 4]
```diff
@@ -430,9 +433,31 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
...
+            self.options.reloadBackgroundOnNext = results[1].reload_background_on_next;
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_form.js:446]: Maybe we could name "results[1]" since it's now being used more? Something like "nextScreenResult" ?

File: addons/survey/static/src/js/survey_session_manage.js [hunk 5]
```diff
@@ -315,54 +318,81 @@ publicWidget.registry.SurveySessionManage = publicWidget.Widget.extend({
             resolveFadeOut();
         });
 
-        var nextQuestionPromise = this._rpc({
-            route: _.str.sprintf('/survey/session/next_question/%s', self.surveyAccessToken),
-            params: {
-                'go_back': goBack,
-            }
-        });
+        var refreshBackground = this.reloadBackgroundOnNext || goBack;
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_session_manage.js:321]: Mmh "refreshBackground" sounds more like a method than a condition.
Not sure what would fit best, "needBackgroundRefresh" maybe?

File: addons/survey/static/src/js/survey_form.js [hunk 6]
```diff
@@ -430,9 +433,31 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
         self.$(selectorsToFadeout.join(',')).fadeOut(this.fadeInOutDelay, function () {
             resolveFadeOut();
         });
+        // Background management - Will the background change between current and next/previous questions ?
+        var refreshBackground = self.options.reloadBackgroundOnNext || options.goBack;
+        if (refreshBackground) {
+            // If we make a transition, hide previous background by setting 'background overlay' opacity = 1
+            $('div#wrapwrap').css("box-shadow", "inset 0 0 0 10000px rgba(255,255,255,1)");
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_form.js:440]: Isn't it possible to make a class out of this?
Would simplify this part and give a name to it since it's used a couple times.
Something like "o_survey_background_transition" ?

File: addons/survey/static/src/js/survey_form.js [hunk 7]
```diff
@@ -500,7 +525,10 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
                 // prevent enter submit if we're on a page description (there is nothing to submit)
                 this.preventEnterSubmit = true;
             }
-
+            // Background management - reset background overlay opacity to 0.7 to discover next background.
+            if (refreshBackground) {
+                $('div#wrapwrap').css("box-shadow", "inset 0 0 0 10000px rgba(255,255,255,0.7)");
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_form.js:530]: And here you would add "o_survey_background_transition o_survey_background_transition_light" ?
And the "_light" rule would only change the rgba in the scss file?

Beware that if you use this approach, you will have to add a couple removeClass as well to handle the back and forth.

File: addons/survey/static/src/js/survey_form.js [hunk 8]
```diff
@@ -1010,6 +1041,25 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
         return false;
     },
 
+   /**
+    * Load the target section background and render it when loaded.
+    *
+    * @private
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_form.js:1047]: Missing `@param {string} imageUrl`

File: addons/survey/static/src/js/survey_form.js [hunk 9]
```diff
@@ -1010,6 +1041,25 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
         return false;
     },
 
+   /**
+    * Load the target section background and render it when loaded.
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_form.js:1045]: Could be nice to explain why we do that (avoid stuttering?)

File: addons/survey/static/src/js/survey_session_manage.js [hunk 10]
```diff
@@ -315,54 +318,81 @@ publicWidget.registry.SurveySessionManage = publicWidget.Widget.extend({
...
+            $('div#wrapwrap').css("box-shadow", "inset 0 0 0 10000px rgba(255,255,255,1)");
+        }
 
         // avoid refreshing results while transitioning
         if (this.resultsRefreshInterval) {
             clearInterval(this.resultsRefreshInterval);
             delete this.resultsRefreshInterval;
         }
 
-        Promise.all([fadeOutPromise, nextQuestionPromise]).then(async function (results) {
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_session_manage.js:331]: On Discord you asked me if we had something to prevent "multi-next" on the session.
I think the answer was "no" because we faded out immediately, so the user did not have the possibility to click multiple times.

Now that you're making that synchronous (we get the next question information and THEN start fading out), maybe it would be nice to prevent the multi-next?
I'm afraid server lags will make the button feel like it's not working during a big session.
At least previously it was fading out directly so you had that feeling of "oh ok it's doing something".

Disabling the button and adding a small spinner somewhere would be nice, I let you see if/how it's doable.

File: addons/survey/static/src/js/survey_session_manage.js [hunk 11]
```diff
@@ -458,6 +488,20 @@ publicWidget.registry.SurveySessionManage = publicWidget.Widget.extend({
         });
     },
 
+    _refreshBackground: function (imageUrl) {
```
Comment (awa-odoo) [inline addons/survey/static/src/js/survey_session_manage.js:491]: I would copy/paste the other docstring here to explain a bit.

File: addons/survey/static/src/scss/survey_form.scss [hunk 12]
```diff
@@ -14,6 +14,10 @@ nav#oe_main_menu_navbar {
 /**********************************************************
                         Common Style
  **********************************************************/
+div#wrapwrap {
+    transition: box-shadow 0.3s ease-in-out;
+}
+
```
Comment (awa-odoo) [inline addons/survey/static/src/scss/survey_form.scss:20]: Ouch.

Can we add the in the survey_form widget?

Like at the start, add a class on the wrapwrap, something like "o_survey_wrapwrap"?
Adding rules on global elements is always a bit scary.

**Review by awa-odoo — COMMENTED** [review]
Hello Davideuh!

This version is looking a lot nicer / simpler than the previous one!
Thank you for sticking with it.

I added some polishing comments and some questions, as always.
Feel free to ping me if something is unclear.

File: addons/survey/static/src/js/survey_session_manage.js (hunk 10 above)
Comment (dbeguin) [reply addons/survey/static/src/js/survey_session_manage.js:331]: Well the issue remains if you use navigation keys. Even if the button is not displayed anymore, you still can "press" go next using arrows. In survey form, I could not break the background. But here I could. Indeed we should probably avoid multi-next, but not only on the button itself. I thought about a debounce on "next question" method.

Comment (awa-odoo) [reply addons/survey/static/src/js/survey_session_manage.js:331]: We could force the display of the little arrow that turns into a spinning loading thing when you press the navigation key as well?

The idea would be to have a "don't worry, it's loading" feeling, and to avoid technical issues of course.

**Review by dbeguin — COMMENTED** [review]
Comment applied. Thanks for your review Rintintin!

File: addons/survey/controllers/main.py [hunk 13]
```diff
@@ -365,16 +383,22 @@ def survey_display_page(self, survey_token, answer_token, **post):
...
             self._prepare_survey_data(access_data['survey_sudo'], answer_sudo, **post))
 
-    @http.route('/survey/get_background_image/<string:survey_token>/<string:answer_token>', type='http', auth="public", website=True, sitemap=False)
-    def survey_get_background(self, survey_token, answer_token):
-        access_data = self._get_access_data(survey_token, answer_token, ensure_token=True)
-        if access_data['validity_code'] is not True:
-            return werkzeug.exceptions.Forbidden()
+    @http.route([
+        '/survey/get_background_image/<string:survey_token>/<int:question_id>',
+        '/survey/get_background_image/<string:survey_token>',
```
Comment (awa-odoo) [inline addons/survey/controllers/main.py:388]: Small side note: would order them the other way around, we usually start with the route with the least amount of arguments and then pile on.

Comment (dbeguin) [reply addons/survey/controllers/main.py:388]: allez, vendu

**Review by tde-banana-odoo — COMMENTED** [review]
Validation, some questions / remarks. Globally a remaining issue is that we have a lot of background image computation in several places, making code hard to follow. Purpose would be to find a way to have self-contained templates (receive a survey and a question and choose what background to display based on it).

**Comment by tde-banana-odoo** [comment]
Closed in favor of https://github.com/odoo/odoo/pull/79156
